enterprise/views.py

```diff
@@ -641,36 +722,80 @@ def post(self, request, enterprise_uuid, course_id):
+        if not selected_course_mode.get('premium') and not user_consent_needed:
+            with transaction.atomic():
+                # Create the Enterprise backend database records for this course
+                # enrollment.
+                EnterpriseCourseEnrollment.objects.get_or_create(
+                    enterprise_customer_user=enterprise_customer_user,
+                    course_id=course_id,
+                )
 
-        with transaction.atomic():
-            # Create the Enterprise backend database records for this course
-            # enrollment, then create the real enrollment in the LMS. Finally,
-            # redirect the user to the courseware page, where they'll be prompted
-            # for data sharing consent if they need to be.
-            ec_user, __ = EnterpriseCustomerUser.objects.get_or_create(
-                user_id=request.user.id,
-                enterprise_customer=enterprise_customer,
+                # In case of Audit course modes enroll the learner directly through
+                # enrollment API client and redirect the learner to dashboard.
+                client = EnrollmentApiClient()
+                client.enroll_user_in_course(request.user.username, course_id, selected_course_mode_name)
+
+            return redirect(LMS_COURSEWARE_URL.format(course_id=course_id))
+
+        # For the audit course modes (audit, honor) or for the premium course
+        # modes (Verified, Prof Ed) where DSC is required, redirect the learner
+        # to course specific DSC with enterprise UUID from where the learner
+        # will be directed to the ecommerce flow after providing DSC.
+        elif user_consent_needed:
+            next_url = '{handle_consent_enrollment_url}?{query_string}'.format(
+                handle_consent_enrollment_url=reverse(
+                    'enterprise_handle_consent_enrollment', args=[enterprise_customer.uuid, course_id]
+                ),
+                query_string=urlencode({'course_mode': selected_course_mode_name})
             )
+            return redirect(
+                '{grant_data_sharing_url}?{params}'.format(
+                    grant_data_sharing_url=reverse('grant_data_sharing_permissions'),
+                    params=urlencode(
+                        {
+                            'next': next_url,
+                            'enterprise_id': enterprise_customer.uuid,
+                            'course_id': course_id,
+                            'enrollment_deferred': True,
+                        }
+                    )
+                )
+            )
+
+        # Learner enrolling in audit course modes (audit, honor) without DSC.
+        else:
             EnterpriseCourseEnrollment.objects.get_or_create(
-                enterprise_customer_user=ec_user,
+                enterprise_customer_user=enterprise_customer_user,
                 course_id=course_id,
             )
```

Comment:
We do not want to create an `EnterpriseCourseEnrollment` record here. This will be done after the learner has paid during fulfillment.